# BIOS60132_Module2
A common task when working with transcriptomic data is the identification of differentially expressed (DE) genes or tags between groups. In this lesson students will learn how to perform biostatistical analysis in the R programming language with the DESeq2 or edgeR packages.

## DE Analysis with DESeq2

### Learning Goals
- Become comfortable working with quantified transcriptomic data (counts)
- Learn what contrasts are possible based on the experimental design
- Be able to perform single or multi factor analysis with DESeq2
- Be able to summarize and visualize the DE analysis results

## DE Analysis with edgeR

### Learning Goals
- Be able to perform pairwise or ANOVA like comparisons comparisons with edgeR
- Become comfortable with setting FDR and LFC thresholds to filter DE results
- Learn methods for identifying unique or shared genes among different sets of DE results
- Learn how to prepare DE results for downstream functional or network analysis

## Prerequisites
- This lesson is designed for students who are unfamiliar with statistical analysis.
- Participants are expected to be comfortable working in the R programming language. 

## Primary Tools
- [R software environment](https://cran.rstudio.com/)
- [RStudio desktop software](https://libcal.library.nd.edu/event/9797081)


